### PR TITLE
Droth 144 tr integration

### PR DESCRIPTION
**Pull request by halied**

### CODE REVIEW

**Comment by halied** [comment]
Looks ok

File: digiroad2-oracle/src/main/scala/fi/liikennevirasto/digiroad2/MassTransitStopService.scala [hunk 1]
```diff
@@ -320,8 +458,24 @@ trait MassTransitStopService extends PointAssetOperations {
     }
   }
 
-  private def getPersistedStopWithPropertiesAndPublishEvent(assetId: Long, roadLinkByLinkId: Long => Option[VVHRoadlink]) = {
+  private def getPersistedStopWithPropertiesAndPublishEvent(assetId: Long, roadLinkByLinkId: Long => Option[VVHRoadlink], createOrUpdateMassTransitStop: TierekisteriMassTransitStop => Unit) = {
     val persistedStop = fetchPointAssets(withId(assetId)).headOption
+    val relevantToTR = isNotVirtualStopAndIsMantainedByELY(Some(persistedStop.get))
+
+    if (relevantToTR) {
+      val (newMTSWithProperties, floatingReason) = persistedStopToMassTransitStopWithProperties(roadLinkByLinkId)(persistedStop.get)
```
Comment (tvenhola) [inline digiroad2-oracle/src/main/scala/fi/liikennevirasto/digiroad2/MassTransitStopService.scala]: These values are not used

File: digiroad2-oracle/src/main/scala/fi/liikennevirasto/digiroad2/MassTransitStopService.scala [hunk 2]
```diff
@@ -320,8 +458,24 @@ trait MassTransitStopService extends PointAssetOperations {
...
+    val relevantToTR = isNotVirtualStopAndIsMantainedByELY(Some(persistedStop.get))
+
+    if (relevantToTR) {
+      val (newMTSWithProperties, floatingReason) = persistedStopToMassTransitStopWithProperties(roadLinkByLinkId)(persistedStop.get)
+      val roadLink = roadLinkByLinkId.apply(persistedStop.get.linkId)
+      val road = roadLink.map(rl => rl.attributes.get("ROADNUMBER")) match {
+        case Some(str) => Try(str.toString.toInt).toOption
+        case _ => None
+      }
+      val (address, roadSide) = geometryTransform.resolveAddressAndLocation(Point(persistedStop.get.lon, persistedStop.get.lat), persistedStop.get.bearing.get)
```
Comment (tvenhola) [inline digiroad2-oracle/src/main/scala/fi/liikennevirasto/digiroad2/MassTransitStopService.scala]: Should we use the road (Option[Int]) formed above to resolve road address? Now road is not used.

File: digiroad2-oracle/src/main/scala/fi/liikennevirasto/digiroad2/MassTransitStopService.scala [hunk 3]
```diff
@@ -320,8 +458,24 @@ trait MassTransitStopService extends PointAssetOperations {
     }
   }
 
-  private def getPersistedStopWithPropertiesAndPublishEvent(assetId: Long, roadLinkByLinkId: Long => Option[VVHRoadlink]) = {
+  private def getPersistedStopWithPropertiesAndPublishEvent(assetId: Long, roadLinkByLinkId: Long => Option[VVHRoadlink], createOrUpdateMassTransitStop: TierekisteriMassTransitStop => Unit) = {
```
Comment (tvenhola) [inline digiroad2-oracle/src/main/scala/fi/liikennevirasto/digiroad2/MassTransitStopService.scala]: Method name "get..." is inconsistent with the save & update Tierekisteri functionality.

File: digiroad2-oracle/src/main/scala/fi/liikennevirasto/digiroad2/MassTransitStopService.scala [hunk 4]
```diff
@@ -454,5 +624,11 @@ trait MassTransitStopService extends PointAssetOperations {
            where id = $id
       """.execute
   }
+
+  def expireMassTransitStop(s: String, id: Long) = {
```
Comment (tvenhola) [inline digiroad2-oracle/src/main/scala/fi/liikennevirasto/digiroad2/MassTransitStopService.scala]: Tierekisteri is not synchronized on this. Should we signal the "valid to date" here as well to Tierekisteri?

File: src/main/scala/fi/liikennevirasto/digiroad2/Digiroad2Api.scala [hunk 5]
```diff
@@ -337,10 +375,16 @@ class Digiroad2Api(val roadLinkService: RoadLinkService,
     assetPropertyService.assetPropertyNames(lang)
   }
 
+  object TierekisteriInternalServerError {
+    def apply(body: Any = Unit, headers: Map[String, String] = Map.empty, reason: String = "") =
+      ActionResult(ResponseStatus(555, reason), body, headers)
```
Comment (tvenhola) [inline src/main/scala/fi/liikennevirasto/digiroad2/Digiroad2Api.scala]: Should we rather use standard error code, such as 502?
"502 Bad Gateway - The server was acting as a gateway or proxy and received an invalid response from the upstream server."

Comment (leirasf) [inline src/main/scala/fi/liikennevirasto/digiroad2/Digiroad2Api.scala]: I believe we have use this different status code to open a different message on client side "Tierekisteri error".

File: UI/src/less/site/global.less [hunk 6]
```diff
@@ -49,13 +49,23 @@ a {
...
+
+  .controlTR {
     display: inline-block;
     padding: 0 10px;
     border-radius: @border-radius-base;
     font-size: @font-size-small;
     line-height: @line-height-modifier * 6;
     color: #fff;
-    background-color: @blue;
+    background-color: #0000ff;
```
Comment (marquesfm) [inline UI/src/less/site/global.less]: Is better put the color "#0000ff" in a variable at .less file like the other color above (@blue):

File: UI/src/view/MassTransitStopForm.js [hunk 7]
```diff
@@ -46,10 +52,19 @@
...
+        selectedMassTransitStopModel.save();
+      }
     });
     var updateStatus = function() {
-      if (selectedMassTransitStopModel.isDirty() && !selectedMassTransitStopModel.requiredPropertiesMissing() && !selectedMassTransitStopModel.hasMixedVirtualAndRealStops()){
+      if (selectedMassTransitStopModel.isDirty() && !selectedMassTransitStopModel.requiredPropertiesMissing() && !selectedMassTransitStopModel.hasMixedVirtualAndRealStops() && !selectedMassTransitStopModel.pikavuoroIsAlone()){
+        element.prop('disabled', false);
+      }
+      else if(selectedMassTransitStopModel.isDirty() && !selectedMassTransitStopModel.requiredPropertiesMissing() && !selectedMassTransitStopModel.hasMixedVirtualAndRealStops() && !selectedMassTransitStopModel.pikavuoroIsAlone()){
+        element.prop('disabled', false);
```
Comment (marquesfm) [inline UI/src/view/MassTransitStopForm.js]: Duplicated "If" should be removed.

File: UI/src/view/MassTransitStopForm.js [hunk 8]
```diff
@@ -547,13 +658,28 @@
...
         var property = event.propertyData;
 
-        if(event.id && property.publicId === 'tietojen_yllapitaja' && (_.find(property.values, function (value) {return value.propertyValue == '2';}))){
+        // Handle form reload after administrator change
+        if (property.publicId === 'tietojen_yllapitaja' && (_.find(property.values, function (value) {return value.propertyValue == '2';}))) {
+          isAdministratorELYKeskus = true;
+          property.propertyType = "single_choice";
+          renderAssetForm();
+        }
+        if (property.publicId === 'tietojen_yllapitaja' && (_.find(property.values, function (value) {return value.propertyValue != '2';}))) {
```
Comment (marquesfm) [inline UI/src/view/MassTransitStopForm.js]: Just put an ELSE for the first IF and it's enough, don't was necessary another if.

File: src/test/scala/fi/liikennevirasto/digiroad2/Digiroad2ApiSpec.scala [hunk 9]
```diff
@@ -1,259 +1,264 @@
...
+  when(mockVVHClient.fetchVVHRoadlinks(Set(1611071l, 1611070l, 1611069l)))
+    .thenReturn(Seq(VVHRoadlink(1611071l, 91,  List(Point(0.0, 0.0), Point(117.318, 0.0)), Municipality, TrafficDirection.UnknownDirection, FeatureClass.AllOthers),
+                    VVHRoadlink(1611070l, 91,  List(Point(117.318, 0.0), Point(127.239, 0.0)), Municipality, TrafficDirection.UnknownDirection, FeatureClass.AllOthers),
+                    VVHRoadlink(1611069l, 91,  List(Point(127.239, 0.0), Point(146.9, 0.0)), Municipality, TrafficDirection.UnknownDirection, FeatureClass.AllOthers)))
+  when(mockVVHClient.fetchVVHRoadlinks(Set(1611374l)))
+    .thenReturn(List(VVHRoadlink(1611374l, 235, Seq(Point(0, 0), Point(120, 0)), Municipality, TrafficDirection.UnknownDirection, FeatureClass.AllOthers)))
+
+
+
+//  1621077551  1611374
```
Comment (marquesfm) [inline src/test/scala/fi/liikennevirasto/digiroad2/Digiroad2ApiSpec.scala]: To remove or is important?

File: UI/src/view/MassTransitStopForm.js (hunk 8 above)
Comment (marquesfm) [inline UI/src/view/MassTransitStopForm.js]: Can be more simplified, for example modified the second IF to ELSE IF, or put the first conditions in a IF and inside evaluate the propertyValue, if is equals to 2 do some actions, ELSE do other actions.

**Comment by marquesfm** [comment]
Seems ok to me, except some errors indicated in the code.

**Pull request merged**